File: src/lexer.cc
#include "lexer.h"

#include <fstream>
#include <unordered_map>

// TODO: some easy optimizations
// could be implemented from https://v8.dev/blog/scanner

std::vector<char> readFile(const std::string& filename) {
	std::ifstream ifs(filename, std::ios::binary | std::ios::ate);
	std::streampos size = ifs.tellg();
	ifs.seekg(0, std::ios::beg);
	std::vector<char> data(size);
	ifs.read(data.data(), size);
	return data;
}

TokenType idToKeyword(const std::string& id) {
	static std::unordered_map<std::string, TokenType> lookup = {
		{"if", TokenType::If},
		{"elif", TokenType::Elif},
		{"else", TokenType::Else},
		{"for", TokenType::For},
		{"while", TokenType::While},
		{"break", TokenType::Break},
		{"switch", TokenType::Switch},
		{"case", TokenType::Case},
		{"and", TokenType::And},
		{"or", TokenType::Or},
		{"type", TokenType::Type},
		{"class", TokenType::Class},
		{"def", TokenType::Def},
		{"public", TokenType::Public},
		{"private", TokenType::Private},
		{"var", TokenType::Var},
		{"let", TokenType::Let},
		{"return", TokenType::Return},
		{"true", TokenType::True},
		{"false", TokenType::False},
		{"module", TokenType::Module},
		{"import", TokenType::Import},
		{"int", TokenType::Int},
		{"int8", TokenType::Int8},
		{"int16", TokenType::Int16},
		{"int32", TokenType::Int32},
		{"int64", TokenType::Int64},
		{"uint", TokenType::Uint},
		{"uint8", TokenType::Uint8},
		{"uint16", TokenType::Uint16},
		{"uint32", TokenType::Uint32},
		{"uint64", TokenType::Uint64},
		{"bool", TokenType::Bool},
		{"float", TokenType::Float},
		{"double", TokenType::Double},
		{"char", TokenType::Char},
		{"string", TokenType::String}
	};
	if (lookup.find(id) != lookup.end()) {
		return lookup[id];
	}
	return TokenType::Invalid;
}

Lexer::Lexer()
	: pos(0) {}

TokenStream* Lexer::lex(const std::string& filename) {
	this->reset();
	this->file = readFile(filename);
	std::vector<Token> tokens;
	while (true) {
		Token t = step();
		tokens.push_back(t);
		if (t.type == TokenType::Eof || t.type == TokenType::Invalid) {
			break;
		}
	}
	return new VectorTokenStream(std::move(tokens));
}

TokenStream* Lexer::lexString(const std::string& source) {
	this->reset();
	this->file = std::vector<char>(source.begin(), source.end());
	std::vector<Token> tokens;
	while (true) {
		Token t = step();
		tokens.push_back(t);
		if (t.type == TokenType::Eof || t.type == TokenType::Invalid) {
			break;
		}
	}
	return new VectorTokenStream(std::move(tokens));
}

void Lexer::reset() {
	this->file.clear();
}

char Lexer::next() {
	return pos < file.size() ? file[pos++] : '\0';
}

void Lexer::back() {
	if (pos < file.size()) {
		pos--;
	}
}

void Lexer::advance() {
	if (pos < file.size()) {
		pos++;
	}
}

char Lexer::peek() {
	return pos+1 < file.size() ? file[pos+1] : '\0';
}

bool Lexer::match(char c) {
	if (peek() == c) {
		advance();
		return true;
	}
	return false;
}

Token Lexer::select(char next, TokenType onMatch, TokenType noMatch) {
	return match(next) ? Token(onMatch) : Token(noMatch);
}

// word lexes identifiers and keywords
Token Lexer::word() {
	int start = pos-1;
	while (true) {
		char c = next();
		if (!std::isalnum(c) && c != '_') {
			back();
			break;
		}
	}
	int end = pos;
	std::string lexeme(file.begin() + start, file.begin() + end);
	TokenType keyword = idToKeyword(lexeme);
	if (keyword != TokenType::Invalid) {
		return Token(keyword);
	}
	return Token(TokenType::Identifier, lexeme);
}

// number lexes number literals
Token Lexer::number() {
	int start = pos-1;
	while (true) {
		char c = next();
		if (!std::isdigit(c)) {
			back();
			break;
		}
	}
	int end = pos;
	std::string lexeme(file.begin() + start, file.begin() + end);
	return Token(TokenType::Number, lexeme);
}

// string lexes string literals
Token Lexer::string() {
	int start = pos; // drop start "
	while (true) {
		char c = next();
		if (c == '"') {
			// consume end "
			break;
		}
	}
	int end = pos-1;
	std::string lexeme(file.begin() + start, file.begin() + end);
	return Token(TokenType::String, lexeme);
}

Token Lexer::comment() {
	int start = pos; // drop the #
	while (next() != '\n') {}
	// drop the newline
	int end = file[pos-2] == '\r' ? pos-3 : pos-2;
	std::string lexeme(file.begin() + start, file.begin() + end);
	return Token(TokenType::Comment, lexeme);
}

void Lexer::skipWspace() {
	while (true) {
		char c = next();
		if (!std::iswspace(c)) {
			back();
			return;
		}
	}
}

Token Lexer::step() {
	skipWspace();

	char c = next();

	if (std::isalpha(c) || c == '_') {
		return word();
	} else if (std::isdigit(c) || (c == '-' && std::isdigit(peek()))) {
		return number();
	}

	switch (c) {
	case '\0':
		if (pos < file.size()) {
			// null character is invalid in source code
			return Token(TokenType::Invalid);
		}
		return Token(TokenType::Eof);

	case '#':
		return comment();

	case '"':
		return string();

	case '+':
		if (match('=')) {
			return Token(TokenType::PlusEq);
		} else if (match('+')) {
			return Token(TokenType::Increment);
		}
		return Token(TokenType::Plus);

	case '-':
		// already checked for negative number literals earlier
		if (match('=')) {
			return Token(TokenType::MinusEq);
		} else if (match('-')) {
			return Token(TokenType::Decrement);
		}
		return Token(TokenType::Minus);

	case '/':
		return select('=', TokenType::DivideEq, TokenType::Divide);

	case '*':
		if (match('*')) {
			return select('=', TokenType::PowerEq, TokenType::Power);
		}
		return select('=', TokenType::StarEq, TokenType::Star);

	case '=':
		return select('=', TokenType::Equal, TokenType::Assign);

	case '!':
		return select('=', TokenType::NotEqual, TokenType::Bang);

	case '<':
		return select('=', TokenType::LessEq, TokenType::Less);

	case '>':
		return select('=', TokenType::GreaterEq, TokenType::Greater);

	case '&':
		return Token(TokenType::BitAnd);

	case '|':
		return Token(TokenType::BitOr);

	case '~':
		return Token(TokenType::BitNegate);

	case '^':
		return Token(TokenType::BitXOR);

	case '.':
		return Token(TokenType::Dot);

	case ',':
		return Token(TokenType::Comma);

	case '(':
		return Token(TokenType::OpenParen);

	case ')':
		return Token(TokenType::CloseParen);

	case '{':
		return Token(TokenType::OpenBrace);

	case '}':
		return Token(TokenType::CloseBrace);

	case '[':
		return Token(TokenType::OpenBrack);

	case ']':
		return Token(TokenType::CloseBrack);

	case ';': // always separates statements
		return Token(TokenType::Sep);
	}

	return Token(TokenType::Invalid);
}

File: ape/lexer.go
package ape

import (
	"os"
	"unicode"

	"github.com/pcen/ape/ape/token"
)

const (
	byteToTokenFactor = 2
)

var (
	stmtEndTokens = map[token.Kind]bool{
		token.Identifier: true,
		token.Integer:    true,
		token.Rational:   true,
		token.String:     true,
		token.True:       true,
		token.False:      true,
		token.Break:      true,
		token.Decrement:  true,
		token.Increment:  true,
		token.Return:     true,
		token.CloseParen: true,
		token.CloseBrace: true,
	}
)

func isalpha(b byte) bool {
	return unicode.IsLetter(rune(b))
}

func isdigit(b byte) bool {
	return unicode.IsDigit(rune(b))
}

func iswspace(b byte) bool {
	return unicode.IsSpace(rune(b))
}

type Lexer interface {
	LexFile(string) []token.Token
	LexString(string) []token.Token
}

func NewLexer() Lexer {
	return &lexer{
		pos: token.Position{
			Line: 1,
			// column index of the next char to read
			// real column number of the most recent char
			Column: 0,
		},
	}
}

type lexer struct {
	buf     []byte
	idx     int
	done    bool
	pos     token.Position
	prevPos token.Position
	tokens  []token.Token
}

func (l *lexer) NewToken(kind token.Kind) token.Token {
	return token.New(kind, l.pos)
}

func (l *lexer) NewLexemeToken(kind token.Kind, lexeme string) token.Token {
	return token.NewLexeme(kind, lexeme, l.pos)
}

func (l *lexer) LexFile(file string) []token.Token {
	var err error
	l.buf, err = os.ReadFile(file)
	if err != nil {
		panic(err)
	}
	return l.lex()
}

func (l *lexer) LexString(source string) []token.Token {
	l.buf = []byte(source)
	return l.lex()
}

func (l *lexer) lex() []token.Token {
	l.tokens = make([]token.Token, 0, len(l.buf)/byteToTokenFactor)
	l.idx = 0
	for {
		tok := l.step()
		l.tokens = append(l.tokens, tok)
		if tok.Kind == token.Eof {
			break
		}
	}
	return l.tokens
}

func (l *lexer) next() (byte, bool) {
	l.done = l.idx == len(l.buf)
	if l.done {
		return 0, false
	}

	b := l.buf[l.idx]
	l.idx++

	if b == '\n' {
		l.prevPos = l.pos
		l.pos.Line++
		l.pos.Column = 0
	} else {
		l.pos.Column++
	}
	return b, true
}

func (l *lexer) back() {
	if l.done {
		return
	}
	l.idx--
	if l.buf[l.idx] == '\n' {
		l.pos = l.prevPos
	} else {
		l.pos.Column--
	}
}

func (l *lexer) peek() byte {
	if l.idx == len(l.buf) {
		return 0
	}
	return l.buf[l.idx]
}

func (l *lexer) match(b byte) bool {
	isMatch := l.peek() == b
	if isMatch {
		l.next() // consume matching token
	}
	return isMatch
}

func (l *lexer) pick(b byte, match token.Kind, noMatch token.Kind) token.Token {
	if l.match(b) {
		return l.NewToken(match)
	}
	return l.NewToken(noMatch)
}

func (l *lexer) shouldInsertSemi(current byte) bool {
	if current != 0 && current != '\n' {
		return false
	}
	if len(l.tokens) == 0 {
		return false // no previous token to check
	}
	return stmtEndTokens[l.tokens[len(l.tokens)-1].Kind]
}

func (l *lexer) skipWhiteSpace() bool {
	for {
		b, ok := l.next()
		if !ok {
			if l.shouldInsertSemi(b) {
				l.tokens = append(l.tokens, l.NewToken(token.Sep))
			}
			return true
		}
		if !iswspace(b) {
			l.back()
			return false
		}
		if l.shouldInsertSemi(b) {
			l.tokens = append(l.tokens, l.NewToken(token.Sep))
		}
	}
}

func (l *lexer) identifier() token.Token {
	start, end := l.idx, 0
	for {
		b, _ := l.next()
		if !isalpha(b) && !isdigit(b) && b != '_' {
			l.back()
			end = l.idx
			break
		}
	}
	lexeme := string(l.buf[start:end])
	if kind, keyword := token.GetKeyword(lexeme); keyword {
		return l.NewToken(kind)
	}
	return l.NewLexemeToken(token.Identifier, lexeme)
}

func (l *lexer) number() token.Token {
	start, end := l.idx, 0
	negative := false
	kind := token.Integer
	for {
		b, _ := l.next()
		if !(isdigit(b) || b == '.' && kind == token.Integer || b == '-' && !negative) {
			l.back()
			end = l.idx
			break
		}
		if b == '.' {
			kind = token.Rational
		}
		negative = b == '-'
	}
	return l.NewLexemeToken(kind, string(l.buf[start:end]))
}

func (l *lexer) comment() token.Token {
	start, end := l.idx, 0
	for {
		b, _ := l.next()
		if b == '\r' {
			l.next()
			end = l.idx - 2
			break
		} else if b == '\n' {
			end = l.idx - 1
			break
		}
	}
	return l.NewLexemeToken(token.Comment, string(l.buf[start:end]))
}

func (l *lexer) str() token.Token {
	start, end := l.idx, 0
	for {
		b, _ := l.next()
		if b == '"' {
			end = l.idx - 1
			break
		}
	}
	return l.NewLexemeToken(token.String, string(l.buf[start:end]))
}

func (l *lexer) step() token.Token {
	atEof := l.skipWhiteSpace()
	if atEof {
		return l.NewToken(token.Eof)
	}
	b, _ := l.next()
	if isalpha(b) || b == '_' {
		// variable or keyword
		l.back()
		return l.identifier()
	}
	if isdigit(b) || (b == '-' && isdigit(l.peek())) {
		// number
		// TODO: parse 0x and b prefixed numbers
		l.back()
		return l.number()
	}

	switch b {
	case '#':
		return l.comment()

	case '"':
		return l.str()

	case '+':
		if l.match('=') {
			return l.NewToken(token.PlusEq)
		} else if l.match('+') {
			return l.NewToken(token.Increment)
		}
		return l.NewToken(token.Plus)

	case '-':
		if l.match('=') {
			return l.NewToken(token.MinusEq)
		} else if l.match('-') {
			return l.NewToken(token.Decrement)
		}
		return l.NewToken(token.Minus)

	case '/':
		return l.pick('=', token.DivideEq, token.Divide)

	case '*':
		if l.match('*') {
			return l.pick('=', token.PowerEq, token.Power)
		}
		return l.pick('=', token.StarEq, token.Star)

	case '=':
		return l.pick('=', token.Equal, token.Assign)

	case '!':
		return l.pick('=', token.NotEqual, token.Bang)

	case '<':
		if l.match('<') {
			return l.NewToken(token.ShiftLeft)
		}
		return l.pick('=', token.LessEq, token.Less)

	case '>':
		if l.match('>') {
			return l.NewToken(token.ShiftRight)
		}
		return l.pick('=', token.GreaterEq, token.Greater)

	case '&':
		return l.NewToken(token.Ampersand)

	case '|':
		return l.NewToken(token.Pipe)

	case '~':
		return l.NewToken(token.Tilde)

	case '^':
		return l.NewToken(token.Caret)

	case '.':
		return l.NewToken(token.Dot)

	case ',':
		return l.NewToken(token.Comma)

	case '(':
		return l.NewToken(token.OpenParen)

	case ')':
		return l.NewToken(token.CloseParen)

	case '{':
		return l.NewToken(token.OpenBrace)

	case '}':
		return l.NewToken(token.CloseBrace)

	case '[':
		return l.NewToken(token.OpenBrack)

	case ']':
		return l.NewToken(token.CloseBrack)

	case ';':
		return l.NewToken(token.Sep)

	}
	panic("invalid byte: " + string(b))
	// return l.NewToken(token.Invalid)
}
